File: scala/atbash-cipher/src/main/scala/Atbash.scala
object Atbash {

  // needed for the tests to compile and run
  def apply() = this

  type Plain = String
  type Cipher = String

  private val CipherGroupSize = 5

  def encode(text: Plain): Cipher = {
    val ungroupedCipherText =
      text.toLowerCase filter (_.isLetterOrDigit) map atbashCipher
    ungroupedCipherText grouped(CipherGroupSize) mkString(" ")
  }

  private val atbashCipher: Map[Char,Char] = {
    val alphabet = ('a' to 'z')
    val cipher = alphabet zip alphabet.reverse toMap

    cipher withDefault identity
  }
}

File: haskell/atbash-cipher/Atbash.hs
module Atbash (encode) where

import Data.Char
import qualified Data.Map as M
import Data.List.Split

type Plain = String
type Cipher = String

encode :: Plain -> Cipher
encode plain =
   group ungroupedCipherText
  where
    ungroupedCipherText = atbashEncode $ filter isLetterOrDigit lowerPlain
    lowerPlain = map toLower plain
    isLetterOrDigit c = isAscii c && (isLetter c || isDigit c)

    atbashEncode = map atbashCipherChar
    atbashCipherChar c = M.findWithDefault c c atbashCipher

    group s = unwords $ chunksOf cipherGroupSize s
    cipherGroupSize = 5

atbashCipher :: M.Map Char Char
atbashCipher =
  M.fromList $ zip alphabet $ reverse alphabet
  where
   alphabet = ['a'..'z']